vowpalwabbit/example.h

```diff
@@ -147,3 +147,6 @@ inline void add_passthrough_feature_magic(example& ec, uint64_t magic, uint64_t
 }
 
 #define add_passthrough_feature(ec, i, x) add_passthrough_feature_magic(ec, __FILE__[0]*483901+__FILE__[1]*3417+__FILE__[2]*8490177, i, x);
+
+typedef v_array<example*> multi_ex;
```

Comment:
Why v_array instead of vector?

Reply:
I experimented with both during development.  

There are pros for v_array.  I can create a sub array without any allocation.   Copy and construction are really cheap.  There are parts of code that pass example collections by value instead of by reference.  This is inexpensive for v_array.  From a performance perspective this is great.

vector<> has other pros.  for example, it has exception safety on modify operations.  Conservative allocation / de-allocation.  Authors of reductions who don't know VW might be less familiar with v_array and therefore more prone to making allocation/de-allocation errors.    

I don't have a clear winner for which is better.

I tried to future proof it behind a somewhat leaky abstraction (typedef) that we can update in the future.  We may want to another type entirely.  For example, we may want to abstract the underlying type to support shared features more directly.

Reply:
My feeling here is that we should generally reserve v_array for instances where it is useful beyond vector since vector is more standard.

Reply:
Makes sense.  Will keep that in mind for future refactoring.